### PR TITLE
fix[JSONSerializer]: fix unserializing of JSON data

### PR DESCRIPTION
**Pull request by spaceone**
1. The json module does not support memoryview() objects to load().

```
>>> import shared_memory_dict
>>> s = shared_memory_dict.SharedMemoryDict('test', 1000, serializer=shared_memory_dict.serializers.JSONSerializer())
>>> s
Fail to load data: TypeError('the JSON object must be str, bytes or bytearray, not memoryview')
{}
>>> s['foo']  = 1
Fail to load data: TypeError('the JSON object must be str, bytes or bytearray, not memoryview')
>>> s
Fail to load data: TypeError('the JSON object must be str, bytes or bytearray, not memoryview')
{}
```

2. The memoryview() data initially contains only null bytes (with the size of the data)

3. If the size decreases, the memory is not overwritten with null bytes.
    Therefore invalid JSON data is part of the memory.
    Always append a null byte to the dumped data and read only until a null
    byte when loading data.

Fixes: d1a564d24ae474d0f1f1d5ff2e16bdf2047919a9

It seems this was never tested - just a broken test case has been added, which I adjusted to more real behavior.
Please also consider my hint https://github.com/luizalabs/shared-memory-dict/commit/77fafc916e62ebd5636a379ee337e3b6158aa1eb#r58780191 - which would have covered this more easily.

### CODE REVIEW

**Comment by cassiobotaro** [comment]
@spaceone I would like to make some suggestions regarding this bug:

Change `_read_memory` to use bytes instead of `memoryview`, keeping the protocol.

```python
return self._serializer.loads(self._memory_block.buf.tobytes())
```
With this change,  should be added an edge case on tests involving null bytes. I suggest creating a constant `NULL_BYTE = b'0\x00'`.

We should add a note on README about this behavior too. Shared memory is initialized with null bytes.

What do you think about initializing memory with empty dict?

```python
    def __init__(
        self,
        name: str,
        size: int,
        *,
        serializer: SharedMemoryDictSerializer = DEFAULT_SERIALIZER,
    ) -> None:
        super().__init__()
        self._serializer = serializer
        self._memory_block = self._get_or_create_memory_block(
            MEMORY_NAME.format(name=name), size
        )
        self._save_memory({})
```

**Comment by spaceone** [comment]
@spaceone OK, `tobytes()` would be good.
initializing with an empty dict is okay, but even better would be to follow a regular dict and initialize it with `dict(*args, **kwargs)`.

**Comment by spaceone** [comment]
> What do you think about initializing memory with empty dict?
> 
> ```diff
>     def __init__(
>         self,
>         name: str,
>         size: int,
>         *,
>         serializer: SharedMemoryDictSerializer = DEFAULT_SERIALIZER,
>     ) -> None:
>         super().__init__()
>         self._serializer = serializer
>         self._memory_block = self._get_or_create_memory_block(
>             MEMORY_NAME.format(name=name), size
>         )
> +       self._save_memory({})
> ```
This would overwrite existing entries, if the memory already exists.

**Comment by cassiobotaro** [comment]
@spaceone Nice catch about `__init__` ! What about?
```python
    def _get_or_create_memory_block(
        self, name: str, size: int
    ) -> SharedMemory:
        try:
            return SharedMemory(name=name)
        except FileNotFoundError:
            smd = SharedMemory(name=name, create=True, size=size)
            smd.buf[:] = b'{}\x00'
            return smd
```

**Comment by spaceone** [comment]
@cassiobotaro Well okay, but `b'{}'\x00` works only for JSON, not for pickle.

**Comment by spaceone** [comment]
@cassiobotaro I pushed:
```
diff --git shared_memory_dict/dict.py shared_memory_dict/dict.py
index c9b002f..340f421 100644
--- shared_memory_dict/dict.py
+++ shared_memory_dict/dict.py
@@ -157,7 +157,9 @@ class SharedMemoryDict:
         try:
             return SharedMemory(name=name)
         except FileNotFoundError:
-            return SharedMemory(name=name, create=True, size=size)
+            shm = SharedMemory(name=name, create=True, size=size)
+            shm.buf[:] = self._serializer.dumps({})
+            return shm
 
     def _save_memory(self, db: Dict[str, Any]) -> None:
         data = self._serializer.dumps(db)

```

File: shared_memory_dict/serializers.py [hunk 1]
```diff
@@ -13,9 +15,12 @@ def loads(self, data: bytes) -> dict:
 
 class JSONSerializer:
     def dumps(self, obj: dict) -> bytes:
-        return json.dumps(obj).encode()
+        return json.dumps(obj).encode() + NULL_BYTE
 
     def loads(self, data: bytes) -> dict:
+        data = bytes(data).split(NULL_BYTE, 1)[0]
+        if not data:
+            return {}
```
Comment (cassiobotaro) [inline shared_memory_dict/serializers.py:23]: As we initialize memory, can we remove these lines?

File: shared_memory_dict/dict.py [hunk 2]
```diff
@@ -157,7 +157,9 @@ def _get_or_create_memory_block(
         try:
             return SharedMemory(name=name)
         except FileNotFoundError:
-            return SharedMemory(name=name, create=True, size=size)
+            shm = SharedMemory(name=name, create=True, size=size)
+            shm.buf[:] = self._serializer.dumps({})
+            return shm
```
Comment (cassiobotaro) [inline shared_memory_dict/dict.py:162]: Can you write a test to check that log will not be printed when memory is fresh and empty?

**Comment by cassiobotaro** [comment]
@spaceone Good catch and nice work on the fix!

**Review by cassiobotaro — APPROVED** [review]

**Review by thyagomininelli — APPROVED** [review]
